In [1]:
s="abc"
print(id(s))

2767991315120


In [5]:
s="hello world"
print(s[-1:-5:-1])

dlro


In [4]:
s="Hello World"
ss=s.replace("l","i",2)
print(ss)

Heiio World


In [7]:
s="abcdabcacdabxyzabcabca"
x=s.find("a")
print("Position : ",x)
for i in range(len(s)):
    z=s.find("a",x+1)
    x=z
    if x==-1:
        break
    print("Position : ",z)

Position :  0
Position :  4
Position :  7
Position :  10
Position :  15
Position :  18
Position :  21


In [28]:
n=int(input("Enter Number :"))
s=int(input("Enter Shift : "))
x=str(n)
y=len(x)
s=s%y
if s==0:
    print(n)
else:
    r=x[s:]+x[:s]
    print(r)

Enter Number : 12345
Enter Shift :  4


51234


# WAP to check if a string is palindrome or not.

In [8]:
s=input("Enter String : ")
r=s[::-1]
if s==r:
    print(s,"is Palindrome String.")
else:
    print(s,"is Not Palindrome String.")

Enter String :  abcd


abcd is Not Palindrome String.


In [7]:
s=input("Enter String : ")
r=s[::-1]
if s==r:
    print(s,"is Palindrome String.")
else:
    print(s,"is Not Palindrome String.")

Enter String :  tat


tat is Palindrome String.


# WAP to Find length of a string in python.

In [11]:
s=input("Enter String : ")
l=len(s)
print("Length of ",s," is ",l)

Enter String :  Tejas Dada


Length of  Tejas Dada  is  10


# WAF to find length of a string in python without using len function.

In [14]:
s=input("Enter String : ")
length(s)
def length(s):
    c=0
    for i in s:
        c+=1
    print("Length of ",s," is ",c)

Enter String :  Tejas Dada


Length of  Tejas Dada  is  10


# WAF that accepts a string and calculate the number of uppercase letters and lowercase letters.

In [16]:
s=input("Enter String :")
uplo(s)
def uplo(s):
    upper=0
    lower=0
    for i in s:
        if i.isupper():
            upper+=1
        if i.islower():
            lower+=1
    print("Upper Case :",upper)
    print("Lower Case :",lower)

Enter String : Mara VaLa


Upper Case : 3
Lower Case : 5


# WAP to demonstrate the negative index in a Tuple

In [19]:
t=eval(input("Enter Tuple :"))
for i in range(-1,-len(t)-1,-1):
    print(i," th index's element is :",t[i])

Enter Tuple : 1,2,3,4


-1  th index's element is : 4
-2  th index's element is : 3
-3  th index's element is : 2
-4  th index's element is : 1


In [21]:
t=(10,20,30,40)
print("Element at -1 index:", t[-1])
print("Element at -2 index:", t[-2])
print("Element at -3 index:", t[-3])
print("Element at -4 index:", t[-4])

Element at -1 index: 40
Element at -2 index: 30
Element at -3 index: 20
Element at -4 index: 10


# WAP to remove I'th character from string in python.

In [35]:
s=input("Enter String :")
i=int(input("Enter the Index : "))
if i<0 or i>len(s):
    print("Invalid index.")
else:
    print(s[:i]+s[i+1:])

Enter String : abcdef
Enter the Index :  3


abcef


In [36]:
s=input("Enter String :")
i=int(input("Enter the Index : "))
if i<0 or i>len(s):
    print("Invalid index.")
else:
    print(s[:i]+s[i+1:])

Enter String : abcdef
Enter the Index :  -8


Invalid index.


# WAP to create a string made of first,middle and last character.

In [3]:
s=input("Enter String :")
le=len(s)
if le<3:
    print("Invaid String.")
else:
    f=s[0]
    l=s[-1]
    m=s[le // 2]
    ns=f+m+l
    print("New String is :",ns)

Enter String : Python


New String is : Phn


In [4]:
s=input("Enter String :")
le=len(s)
if le<3:
    print("Invaid String.")
else:
    f=s[0]
    l=s[-1]
    m=s[le // 2]
    ns=f+m+l
    print("New String is :",ns)

Enter String : nayan


New String is : nyn


# WAP to find all occuences of a sub string in a given string by ignoring the case

In [9]:
s=input("Enter the String :")
sub=input("Enter The Sub String :")
s=s.lower()
sub=sub.lower()
c=s.count(sub)
print(c)

Enter the String : ABCDabAbCdsbAbCD
Enter The Sub String : abcD


3


# WAP to calculate the sum and average of the digits present in a string.

In [15]:
s=input("Enter the String :")
sum=0
c=0
for i in s :
    if i.isdigit() :
        sum+=int(i)
        c+=1
print("Sum : ",sum)
avg=sum/c
print("Average :",avg)

Enter the String : Pyt6th4o9n3


Sum :  22
Average : 5.5


# WAP to reverse a given string

In [4]:
# String function

s="Python"
print(s[::-1])

nohtyP


In [5]:
# WithOut String Function

s="Python"
n=""
for i in range(-1,-len(s)-1,-1):
    n+=s[i]
print(n)

nohtyP


# WAP to print even length words in a string.

In [2]:
s="Python is very easy languvage"
n=s.split(" ")
for i in n:
    if len(i)%2==0:
        print(i)

Python
is
very
easy


# WAP to  Uppercase Half String from the given string.

In [12]:
s="python is very easy"
l=len(s)
half=l//2
upper=s[:half].upper()
lower=s[half:].lower()
print(upper,end="")
print(lower)

PYTHON IS very easy


In [11]:
s="python"
l=len(s)
half=l//2
new=s[:half].upper()+s[half:].lower()
print(new)

PYThon


# WAP to capitalize the first and last character of each word in a string

In [19]:
s="python is a very easy"
n=s.split(" ")
new=[]
for i in n:
    if len(i)==1:
        new.append(i.upper())
    else:
        new.append(i[0].upper()+i[1:-1:]+i[-1].upper())
for j in new:
    print(j,end=" ")

PythoN IS A VerY EasY 

In [21]:
s="python is a very easy"
n=s.split(" ")
new=[]
for i in n:
    if len(i)==1:
        new.append(i.upper())
    else:
        new.append(i[0].upper()+i[1:-1:]+i[-1].upper())
r=" ".join(new)
print(r)

PythoN IS A VerY EasY


# WAP to Create a string made of the middle three characters

In [24]:
s="python"
if len(s)<3:
    print("Invalid length")
else: 
    middle=len(s)//2
    new=s[middle-1]+s[middle]+s[middle+1]
    print(new)

tho


In [28]:
s="python is very easy"
if len(s)<3:
    print("Invalid length")
else: 
    middle=len(s)//2
    new=s[middle-1:middle+2]
    print(new)

s v


# WAP to check if two strings are balanced. For example, strings s1 and s2 are balanced if all the characters in the s1 are present in s2. The character’s position doesn’t matter.

In [4]:
a=input("Enter String 1 :")
b=input("Enter String 2 :")
c=0
for i in a :
    if i in b:
        c+=1
if c==len(a):
    print(a," and ",b," are balanced.")
else:
    print(a," and ",b," are Not balanced.")

Enter String 1 : abc
Enter String 2 : xyajdbkdc


abc  and  xyajdbkdc  are balanced.


In [3]:
a=input("Enter String 1 :")
b=input("Enter String 2 :")
balance=True
for i in a :
    if i not in b:
        balanced=Falsepython-is-very-easy
if balanced:
    print(a," and ",b," are balanced.")
else:
    print(a," and ",b," are Notpython-is-very-easy balanced.")

Enter String 1 : csd
Enter String 2 : nksjfl


csd  and  nksjfl  are Not balanced.


# WAP to Split a string on hyphens.

In [5]:
s=input("Enter the String : ")
new=s.split("-")
for i in new:
    print(i)

Enter the String :  python-is-very-easy


python
is
very
easy


# WAP to print maximum and minimum elements in given Tuple.

In [13]:
s=eval(input("Enter Data:"))
print("Max :",max(s))
print("Min :",min(s))

Enter data: 10,20,30,40


Max : 40
Min : 10


In [19]:
s=eval(input("Enter Data:"))
max=0
for i in s:
    if i>max:
        max=i
min=max
for j in s:
    if j<min:
        min=j
print("Max :",max)
print("Min :",min)

Enter Data: 6,8,5,6,9,4,3,1,7,6


Max : 9
Min : 1


# WAP to print even numbers from given Tuple.

In [21]:
t=eval(input("Enter tuple :"))
print("Even Number's are :",end="  ")
for i in t:
    if i%2==0:
        print(i,end=" ")

Enter tuple : 5,6,4,3,8,5,6,7,9,2,5,45,65,52


Even Number's are :  6 4 8 6 2 52 

# WAP to print sum of even numbers and sum of odd numbers from elements given in tuple.

In [22]:
t=eval(input("Enter tuple :"))
sum_even=0
sum_odd=0
for i in t:
    if i%2==0:
        sum_even+=i
    else:
        sum_odd+=i
print("Sum of even Number's is ",sum_even)
print("Sum of odd Number's is ",sum_odd)

Enter tuple : 6,5,7,5,6,3,2,4,1,9,8


Sum of even Number's is  26
Sum of odd Number's is  30


# WAP Write a Python program using function to shift the decimal digits n places to the left, wrapping the extra digits around. If shift > the number of digits of n, then reverse the string.Note: Function will take two parameters: 1. The number 2. How much shift user want.
#### Example:
#### Input: n=12345 shift=1
#### Output: Result=23451
#### Input: n=12345 shift=3
#### Output: Result=45123
#### Input: n=12345 shift=5
#### Output: Result=12345
#### Input: n=12345 shift=6
#### Output: Result=54321

In [36]:
n=input("Enter Number :")
s=int(input("Enter Shift :"))
def shift(n,s):
    if s>len(n):
        print(n[::-1])
    elif s==len(n):
        print(n)
    else:
        print(n[s::1]+n[s-1::-1])
shift(n,s)

Enter Number : 12345
Enter Shift : 1


23451


In [37]:
n=input("Enter Number :")
s=int(input("Enter Shift :"))
def shift(n,s):
    if s>len(n):
        print(n[::-1])
    elif s==len(n):
        print(n)
    else:
        print(n[s::1]+n[s-1::-1])
shift(n,s)

Enter Number : 12345
Enter Shift : 4


54321


In [38]:
n=input("Enter Number :")
s=int(input("Enter Shift :"))
def shift(n,s):
    if s>len(n):
        print(n[::-1])
    elif s==len(n):
        print(n)
    else:
        print(n[s::1]+n[s-1::-1])
shift(n,s)

Enter Number : 12345
Enter Shift : 5


12345


In [39]:
n=input("Enter Number :")
s=int(input("Enter Shift :"))
def shift(n,s):
    if s>len(n):
        print(n[::-1])
    elif s==len(n):
        print(n)
    else:
        print(n[s::1]+n[s-1::-1])
shift(n,s)

Enter Number : 12345
Enter Shift : 7


54321


# WAP Write a Python programme that accepts a string and calculate the number of uppercase letters, lowercase letters and number of digits.
#### For example,
#### Input: Hello Pyth@n is 100% easy
#### Output:
#### Uppercase letters : 2
#### Lowercase letters : 14
#### Digits : 3

In [40]:
s=input("Enter String :")
digit=0
upper=0
lower=0
for i in s:
    if i.isupper():
        upper+=1
    elif i.islower():
        lower+=1
    elif i.isdigit():
        digit+=1
    else:
        pass
print("Uppercas Letters :",upper)
print("Lowercase Letters :",lower)
print("Digit's :",digit)

Enter String : Hello Pyth@n is 100% easy


Uppercas Letters : 2
Lowercase Letters : 14
Digit's : 3


# WAP to check the validity of a Password.
# Primary conditions for password validation:
# 1. Minimum 8 characters.
# 2. The alphabet must be between [a-z]
# 3. At least one alphabet should be of Upper Case [A-Z]
# 4. At least 1 number or digit between [0-9]
# 5. At least 1 character from [ _ or @ or $]
#### Examples:
#### Input: Ram@_f1234
#### Output: Valid Password
#### Input: Rama_fo$ab
#### Output: Invalid Password
#### Explanation: Number is missing
#### Input: Rama#fo9c
#### Output: Invalid Password
#### Explanation: Must consist from _ or @ or $

In [55]:
p=input("Enter Password :")
upper=0
lower=0
digit=0
sp=0
l=len(p)
if l>=8:
    for i in p:
        if i.isupper():
            upper+=1
        elif i.islower():
            lower+=1
        elif i.isdigit():
            digit+=1
        elif i=="_" or i=="@" or i=="$":
            sp+=1
        else:
            pass
    total=upper+lower+digit+sp
    if upper>=1:
        if digit>=1:
            if sp>=1:
                print("Password is valid.")
            else:
                print("Password is invalid.")
                print("Password must have from _,@,$.")
        else:
            print("Password is invalid.")
            print("Password must have at least 1 digit.")
    else:
        print("Password is invalid.")
        print("Password must have at least 1 Uppercase.")
else:
    print("Password is invalid.")
    print("Length must be minimum 8 digit.")

Enter Password : Ram@_f1234


Password is valid.


In [53]:
p=input("Enter Password :")
upper=0
lower=0
digit=0
sp=0
l=len(p)
if l>=8:
    for i in p:
        if i.isupper():
            upper+=1
        elif i.islower():
            lower+=1
        elif i.isdigit():
            digit+=1
        elif i=="_" or i=="@" or i=="$":
            sp+=1
        else:
            pass
    total=upper+lower+digit+sp
    if upper>=1:
        if digit>=1:
            if sp>=1:    
                print("Password is valid.")
            else:
                print("Password is invalid.")
                print("Password must have from _,@,$.")
        else:
            print("Password is invalid.")
            print("Password must have at least 1 digit.")
    else:
        print("Password is invalid.")
        print("Password must have at least 1 Uppercase.")
else:
    print("Password is invalid.")
    print("Length must be minimum 8 digit.")

Enter Password : Rama_fo$ab


Password is invalid.
Password must have at least 1 digit.


In [52]:
p=input("Enter Password :")
upper=0
lower=0
digit=0
sp=0
l=len(p)
if l>=8:
    for i in p:
        if i.isupper():
            upper+=1
        elif i.islower():
            lower+=1
        elif i.isdigit():
            digit+=1
        elif i=="_" or i=="@" or i=="$":
            sp+=1
        else:
            pass
    total=upper+lower+digit+sp
    if upper>=1:
        if digit>=1:
            if sp>=1:
                print("Password is valid.")
            else:
                print("Password is invalid.")
                print("Password must have from _,@,$.")
        else:
            print("Password is invalid.")
            print("Password must have at least 1 digit.")
    else:
        print("Password is invalid.")
        print("Password must have at least 1 Uppercase.")
else:
    print("Password is invalid.")
    print("Length must be minimum 8 digit.")

Enter Password : Rama#fo9c


Password is invalid.
Password must have from _,@,$.


# WAP to return another string similar to the input string, but with its case inverted. For example, input of “Mr. Ed” will result in “mR. eD” as the output string.
# Note: Use of built in swapcase function is prohibited.

In [59]:
s=input("Enter String :")
new=""
for i in s:
    if i.islower():
        new+=i.upper()
    elif i.isupper():
        new+=i.lower()
    else:
        new+=i
print(new)

Enter String : Mr. Ed


mR. eD


In [60]:
s=input("Enter String :")
new=""
for i in s:
    if i.islower():
        new+=i.upper()
    elif i.isupper():
        new+=i.lower()
    else:
        new+=i
print(new)

Enter String : TejaS DAda


tEJAs daDA


# WAP to check if two strings are balanced. For example, strings s1 and s2 are balanced if all the characters in the s1 are present in s2 and length of s1 & s2 should be same. The character’s position doesn’t matter.
#### Example :
#### s1 = hello
#### s2 = olleh
#### Balanced

In [6]:
s1=input("Enter String 1 :")
s2=input("Enter String 2 :")
c=0
if len(s1)==len(s2):
    for i in s1:
        if i in s2:
            c+=1
else:
    print("String is Unbalanced.")
    print("Length of both String can't match.")
if c==len(s1):
    print("String is Balanced.")

Enter String 1 : hello
Enter String 2 : olleh


String is Balanced.


In [5]:
s1=input("Enter String 1 :")
s2=input("Enter String 2 :")
c=0
if len(s1)==len(s2):
    for i in s1:
        if i in s2:
            c+=1
else:
    print("String is Unbalanced.")
    print("Length of both String can't match.")
if c==len(s1):
    print("String is Balanced.")

Enter String 1 : Hello World
Enter String 2 : HelloWo rlds


String is Unbalanced.
Length of both String can't match.


In [4]:
s1=input("Enter String 1 :")
s2=input("Enter String 2 :")
c=0
if len(s1)==len(s2):
    for i in s1:
        if i in s2:
            c+=1
else:
    print("String is Unbalanced.")
    print("Length of both String can't match.")
if c==len(s1):
    print("String is Balanced.")

Enter String 1 : hii sir
Enter String 2 : hiisi r


String is Balanced.


#  Write a Python program to create a Caesar encryption.
# Note: In cryptography, a Caesar cipher, also known as Caesar's cipher, the shift cipher, Caesar's code or Caesar shift, is one of the simplest and most widely known encryption techniques. It is a type of substitution cipher in which each letter in the plaintext is replaced by a letter some fixed number of positions down the alphabet. For example, with a right shift of 3, A would be replaced by D, E would become H, and so on. The method is named after Julius Caesar, who used it in his private correspondence.
#### For Example:
#### Input Text  : LJIET ENG
#### Shift : 3
#### Cipher:  OMLHW HQJ

# WAP to print the even numbers from a given list.

In [4]:
l=eval(input("Enter List :"))
print("Even Numbers : ")
for i in l:
    if i%2==0:
        print(i,end=" ")

Enter List : [5,6,9,2,3,4,8,6,45,75,87,68,1,0,9]


Even Numbers : 
6 2 4 8 6 68 0 

# WAP to print the largest even number in a list.

In [5]:
l=eval(input("Enter List :"))
maxi=[]
for i in l:
    if i%2==0:
        maxi.append(i)
print("Large even Number in List : ",max(maxi))

Enter List : [8,6,5,2,3,4,9,7,6,58,45,32,10,59]


Large even Number in List :  58


# WAP to print the largest odd number in a list.

In [6]:
l=eval(input("Enter List :"))
mini=[]
for i in l:
    if i%2!=0:
        mini.append(i)
print("Large odd Number in List : ",max(mini))

Enter List : [8,6,5,2,3,4,9,7,6,58,45,32,10,59]


Large odd Number in List :  59


# WAP to swap first and last element of the list. 

In [7]:
l=eval(input("Enter List : "))
a=l[0]
l[0]=l[-1]
l[-1]=a
print("List After Swaped :",l)

Enter List :  [5,6,8,2,1,6,2,4,7]


List After Swaped : [7, 6, 8, 2, 1, 6, 2, 4, 5]


# WAP to find the sum of all the elements in the list. 

In [8]:
l=eval(input("Enter List :"))
sum=0
for i in l:
    sum+=i
print("Sum Of The list is :",sum)

Enter List : [5,2,3,6,4]


Sum Of The list is : 20


# WAF to sum all the numbers in a list.

In [9]:
def sum_list(l):
    sum=0
    for i in l:
        sum+=i
    return sum
l=eval(input("Enter List :"))
print("Sum Of The List is :",sum_list(l))

Enter List : [5,2,3,6,4]


Sum Of The List is : 20


# WAP of Reversing a List.

In [11]:
l=eval(input("Enter List :"))
l.reverse()
print("List After Reverce :",l)

Enter List : [6,5,8,2,3,9,4]


List After Reverce : [4, 9, 3, 2, 8, 5, 6]


# WAP to Merging two Dictionaries.

In [18]:
d1={1:"Nobita",2:"Sizuka",3:"Doremon"}
d2={"Yogi":4,"Motu":5,"Patlu":6}
d1.update(d2)
print(d1)

{1: 'Nobita', 2: 'Sizuka', 3: 'Doremon', 'Yogi': 4, 'Motu': 5, 'Patlu': 6}


# WAP Given a list of elements, write a python program to perform grouping of similar elements, as different key-value list in dictionary. Print the dictionary sorted in descending order of frequency of the elements.
## Note: To perform the sorting, use the sorted function by converting the dictionary into a list of tuples. After sorting, convert the list of tuples back into a dictionary and print it.
#### Input : test_list = [4, 6, 6, 4, 2, 2, 4, 8, 5, 8] 
#### Output : {4: [4, 4, 4], 6: [6, 6], 2: [2, 2], 8: [8, 8], 5: [5]} 
## Explanation : Similar items grouped together on occurrences. 
#### Input : test_list = [7, 7, 7, 7] 
#### Output : {7 : [7, 7, 7, 7]} 
## Explanation : Similar items grouped together on occurrences.

In [32]:
l=[4,6,6,4,2,2,4,8,5,8]
d={}
for i in l:
    if i in d:
        d[i].append(i)
    else:
        d[i]=[i]
nl=sorted(d.items(),key=lambda x:len(x[1]),reverse=True)
print("Dict : ",dict(nl))

Dict :  {4: [4, 4, 4], 6: [6, 6], 2: [2, 2], 8: [8, 8], 5: [5]}


# Write a Program to Print Longest Common Prefix from a given list of strings. The longest common prefix for list of strings is the common prefix (starting of string) between all strings. For example, in the given list [“apple”, “ape”, “zebra”], there is no common prefix because the 2 most dissimilar strings of the list “ape” and “zebra” do not share any starting characters. If there is no common prefix between all strings in the list than return -1.
#### For example,
#### Input list: ["lessonplan", "lesson","lees", "length"]
#### The longest Common Prefix is: le
#### Input list: ["python","pythonprogramming","pythonlist"]
#### The longest Common Prefix is: python
#### Input list: ["lessonplan", "lesson","ees", "length"]
#### The longest Common Prefix is: -1

In [1]:
def longest_common_prefix(words):
    if not words:
        return -1

    # Step 1: Sort words (smallest difference will appear first)
    words.sort()

    # Step 2: Compare first and last word only
    first = words[0]
    last = words[-1]

    prefix = ""

    for i in range(min(len(first), len(last))):
        if first[i] == last[i]:
            prefix += first[i]
        else:
            break

    return prefix if prefix else -1


# -------------------------
# Testing
# -------------------------

print(longest_common_prefix(["lessonplan", "lesson","lees", "length"]))  
print(longest_common_prefix(["python","pythonprogramming","pythonlist"]))
print(longest_common_prefix(["lessonplan", "esson","lees", "lngth"]))


le
python
-1
